File: 2018/day02/python/main.py
def part1(input):
    two = three = 0

    for line in input:
        counts = {}
        for char in line: counts[char] = counts.get(char, 0) + 1
        if 2 in counts.values(): two += 1
        if 3 in counts.values(): three += 1

    return two * three

def part2(input):
    for i in range(len(input)):
        for j in range(i + 1, len(input)):
            diff = [a != b for a, b in zip(input[i], input[j])]
            if sum(diff) == 1:
                return ''.join([input[i][k] for k in range(len(input[i])) if not diff[k]])

input = open('../input.txt').read().splitlines()

print(part1(input))
print(part2(input))

File: 2018/day02/rust/src/main.rs
use std::{
    collections::HashMap,
    fs::File,
    io::{BufRead, BufReader},
    path::Path,
};

fn main() {
    let path = Path::new("../input.txt");
    let file = File::open(&path).expect("could not open file");
    let reader = BufReader::new(file);
    let input: Vec<String> = reader
        .lines()
        .map(|line| line.expect("could not read line"))
        .collect();

    let mut total = (0, 0);
    let mut common: Option<String> = None;

    for i in 0..input.len() {
        let mut chars: HashMap<char, i32> = HashMap::new();

        for char in input[i].chars() {
            *chars.entry(char).or_insert(0) += 1;
        }

        if chars.values().any(|&v| v == 2) {
            total.0 += 1;
        }

        if chars.values().any(|&v| v == 3) {
            total.1 += 1;
        }

        if common == None {
            for j in i + 1..input.len() {
                let diff: Vec<bool> = input[i]
                    .chars()
                    .zip(input[j].chars())
                    .map(|(a, b)| a == b)
                    .collect();

                if diff.iter().filter(|&&d| !d).count() == 1 {
                    common = Some(
                        input[i]
                            .chars()
                            .zip(diff.iter())
                            .filter_map(|(c, &d)| if d { Some(c) } else { None })
                            .collect(),
                    );
                }
            }
        }
    }

    println!("{}\n{}", total.0 * total.1, common.unwrap());
}
